File: python/src/acedatacloud/_runtime/transport.py
"""HTTP transport layer for AceDataCloud SDK."""

from __future__ import annotations

import inspect
import os
import random
import time
from collections.abc import Iterator
from typing import Any

import httpx

from acedatacloud._runtime.errors import (
    APIError,
    AuthenticationError,
    InsufficientBalanceError,
    ModerationError,
    RateLimitError,
    ResourceDisabledError,
    TimeoutError,
    TokenMismatchError,
    TransportError,
    ValidationError,
)
from acedatacloud._runtime.payment import (
    PaymentHandler,
    SyncPaymentHandler,
)

_ERROR_CODE_MAP = {
    "invalid_token": AuthenticationError,
    "token_expired": AuthenticationError,
    "no_token": AuthenticationError,
    "token_mismatched": TokenMismatchError,
    "used_up": InsufficientBalanceError,
    "disabled": ResourceDisabledError,
    "too_many_requests": RateLimitError,
    "bad_request": ValidationError,
}

_RETRY_STATUS_CODES = {408, 409, 429, 500, 502, 503, 504}


def _map_error(status_code: int, body: dict[str, Any]) -> APIError:
    """Map an API error response to the appropriate exception class."""
    error_data = body.get("error", {})
    code = error_data.get("code", "")
    message = error_data.get("message", "")
    trace_id = body.get("trace_id")

    exc_class = _ERROR_CODE_MAP.get(code)
    if exc_class is None:
        if status_code == 403:
            exc_class = ModerationError
        elif status_code == 401:
            exc_class = AuthenticationError
        elif status_code == 429:
            exc_class = RateLimitError
        elif status_code == 400:
            exc_class = ValidationError
        else:
            exc_class = APIError

    return exc_class(
        message=message,
        status_code=status_code,
        code=code,
        trace_id=trace_id,
        body=body,
    )


def _should_retry(status_code: int) -> bool:
    return status_code in _RETRY_STATUS_CODES


def _backoff_delay(attempt: int) -> float:
    base = min(2**attempt, 8)
    jitter = random.random() * 0.5  # noqa: S311
    return base + jitter


class SyncTransport:
    """Synchronous HTTP transport using httpx."""

    def __init__(
        self,
        api_token: str | None,
        base_url: str,
        platform_base_url: str,
        timeout: float,
        max_retries: int,
        extra_headers: dict[str, str],
        payment_handler: SyncPaymentHandler | None = None,
    ) -> None:
        token = api_token or os.environ.get("ACEDATACLOUD_API_TOKEN", "")
        if not token and payment_handler is None:
            raise AuthenticationError(
                message=(
                    "api_token is required (or provide a payment_handler). "
                    "Pass it to the client or set ACEDATACLOUD_API_TOKEN."
                ),
                status_code=0,
                code="no_token",
            )
        self._base_url = base_url.rstrip("/")
        self._platform_base_url = platform_base_url.rstrip("/")
        self._timeout = timeout
        self._max_retries = max_retries
        self._payment_handler = payment_handler
        headers: dict[str, str] = {
            "accept": "application/json",
            "content-type": "application/json",
            "user-agent": "acedatacloud-python/0.1.0",
            **extra_headers,
        }
        if token:
            headers["authorization"] = f"Bearer {token}"
        self._headers = headers
        self._client = httpx.Client(timeout=timeout)

    def request(
        self,
        method: str,
        path: str,
        *,
        json: dict[str, Any] | None = None,
        params: dict[str, Any] | None = None,
        platform: bool = False,
        timeout: float | None = None,
        extra_headers: dict[str, str] | None = None,
    ) -> dict[str, Any]:
        base = self._platform_base_url if platform else self._base_url
        url = f"{base}{path}"
        headers = {**self._headers, **(extra_headers or {})}
        extra_auth_headers: dict[str, str] = {}
        payment_attempted = False

        last_exc: Exception | None = None
        for attempt in range(self._max_retries + 1):
            try:
                resp = self._client.request(
                    method,
                    url,
                    json=json,
                    params=params,
                    headers={**headers, **extra_auth_headers},
                    timeout=timeout or self._timeout,
                )
            except httpx.TimeoutException as exc:
                last_exc = TimeoutError(
                    message=f"Request timed out: {exc}",
                    status_code=0,
                    code="timeout",
                )
                if attempt < self._max_retries:
                    time.sleep(_backoff_delay(attempt))
                    continue
                raise last_exc from exc
            except httpx.TransportError as exc:
                last_exc = TransportError(str(exc))
                if attempt < self._max_retries:
                    time.sleep(_backoff_delay(attempt))
                    continue
                raise last_exc from exc

            if (
                resp.status_code == 402
                and self._payment_handler is not None
                and not payment_attempted
            ):
                try:
                    body = resp.json()
                except Exception as exc:
                    raise _map_error(
                        402,
                        {"error": {"code": "invalid_402", "message": resp.text}},
                    ) from exc
                accepts = body.get("accepts") or []
                if not accepts:
                    raise _map_error(
                        402,
                        {"error": {"code": "invalid_402", "message": "No payment requirements"}},
                    )
                result = self._payment_handler(
                    {"url": url, "method": method, "body": json, "accepts": accepts}
                )
                extra_auth_headers.update(result.get("headers", {}))
                payment_attempted = True
                continue

            if resp.status_code >= 400:
                try:
                    body = resp.json()
                except Exception:
                    body = {"error": {"code": "unknown", "message": resp.text}}

                if _should_retry(resp.status_code) and attempt < self._max_retries:
                    time.sleep(_backoff_delay(attempt))
                    continue
                raise _map_error(resp.status_code, body)

            return resp.json()

        raise last_exc or TransportError("Request failed after retries")

    def request_stream(
        self,
        method: str,
        path: str,
        *,
        json: dict[str, Any] | None = None,
        timeout: float | None = None,
        extra_headers: dict[str, str] | None = None,
    ) -> Iterator[str]:
        """Send a request and yield SSE data lines."""
        base = self._base_url
        url = f"{base}{path}"
        headers = {
            **self._headers,
            "accept": "text/event-stream",
            **(extra_headers or {}),
        }

        with self._client.stream(
            method,
            url,
            json=json,
            headers=headers,
            timeout=timeout or self._timeout,
        ) as resp:
            if resp.status_code >= 400:
                body_bytes = resp.read()
                try:
                    import json as _json

                    body = _json.loads(body_bytes)
                except Exception:
                    body = {"error": {"code": "unknown", "message": body_bytes.decode(errors="replace")}}
                raise _map_error(resp.status_code, body)

            for line in resp.iter_lines():
                if line.startswith("data: "):
                    data = line[6:]
                    if data == "[DONE]":
                        return
                    yield data

    def upload(
        self,
        path: str,
        file_data: bytes,
        filename: str,
        *,
        timeout: float | None = None,
    ) -> dict[str, Any]:
        url = f"{self._platform_base_url}{path}"
        headers = {k: v for k, v in self._headers.items() if k != "content-type"}

        resp = self._client.post(
            url,
            files={"file": (filename, file_data)},
            headers=headers,
            timeout=timeout or self._timeout,
        )
        if resp.status_code >= 400:
            try:
                body = resp.json()
            except Exception:
                body = {"error": {"code": "unknown", "message": resp.text}}
            raise _map_error(resp.status_code, body)
        return resp.json()

    def close(self) -> None:
        self._client.close()


class AsyncTransport:
    """Asynchronous HTTP transport using httpx."""

    def __init__(
        self,
        api_token: str | None,
        base_url: str,
        platform_base_url: str,
        timeout: float,
        max_retries: int,
        extra_headers: dict[str, str],
        payment_handler: PaymentHandler | None = None,
    ) -> None:

        token = api_token or os.environ.get("ACEDATACLOUD_API_TOKEN", "")
        if not token and payment_handler is None:
            raise AuthenticationError(
                message=(
                    "api_token is required (or provide a payment_handler). "
                    "Pass it to the client or set ACEDATACLOUD_API_TOKEN."
                ),
                status_code=0,
                code="no_token",
            )
        self._base_url = base_url.rstrip("/")
        self._platform_base_url = platform_base_url.rstrip("/")
        self._timeout = timeout
        self._max_retries = max_retries
        self._payment_handler = payment_handler
        headers: dict[str, str] = {
            "accept": "application/json",
            "content-type": "application/json",
            "user-agent": "acedatacloud-python/0.1.0",
            **extra_headers,
        }
        if token:
            headers["authorization"] = f"Bearer {token}"
        self._headers = headers
        self._client = httpx.AsyncClient(timeout=timeout)

    async def request(
        self,
        method: str,
        path: str,
        *,
        json: dict[str, Any] | None = None,
        params: dict[str, Any] | None = None,
        platform: bool = False,
        timeout: float | None = None,
        extra_headers: dict[str, str] | None = None,
    ) -> dict[str, Any]:
        import asyncio

        base = self._platform_base_url if platform else self._base_url
        url = f"{base}{path}"
        headers = {**self._headers, **(extra_headers or {})}
        extra_auth_headers: dict[str, str] = {}
        payment_attempted = False

        last_exc: Exception | None = None
        for attempt in range(self._max_retries + 1):
            try:
                resp = await self._client.request(
                    method,
                    url,
                    json=json,
                    params=params,
                    headers={**headers, **extra_auth_headers},
                    timeout=timeout or self._timeout,
                )
            except httpx.TimeoutException as exc:
                last_exc = TimeoutError(
                    message=f"Request timed out: {exc}",
                    status_code=0,
                    code="timeout",
                )
                if attempt < self._max_retries:
                    await asyncio.sleep(_backoff_delay(attempt))
                    continue
                raise last_exc from exc
            except httpx.TransportError as exc:
                last_exc = TransportError(str(exc))
                if attempt < self._max_retries:
                    await asyncio.sleep(_backoff_delay(attempt))
                    continue
                raise last_exc from exc

            if (
                resp.status_code == 402
                and self._payment_handler is not None
                and not payment_attempted
            ):
                try:
                    body = resp.json()
                except Exception as exc:
                    raise _map_error(
                        402,
                        {"error": {"code": "invalid_402", "message": resp.text}},
                    ) from exc
                accepts = body.get("accepts") or []
                if not accepts:
                    raise _map_error(
                        402,
                        {"error": {"code": "invalid_402", "message": "No payment requirements"}},
                    )
                handler_result = self._payment_handler(
                    {"url": url, "method": method, "body": json, "accepts": accepts}
                )
                if inspect.isawaitable(handler_result):
                    handler_result = await handler_result
                extra_auth_headers.update(handler_result.get("headers", {}))
                payment_attempted = True
                continue

            if resp.status_code >= 400:
                try:
                    body = resp.json()
                except Exception:
                    body = {"error": {"code": "unknown", "message": resp.text}}

                if _should_retry(resp.status_code) and attempt < self._max_retries:
                    await asyncio.sleep(_backoff_delay(attempt))
                    continue
                raise _map_error(resp.status_code, body)

            return resp.json()

        raise last_exc or TransportError("Request failed after retries")

    async def request_stream(
        self,
        method: str,
        path: str,
        *,
        json: dict[str, Any] | None = None,
        timeout: float | None = None,
        extra_headers: dict[str, str] | None = None,
    ):
        """Send a request and yield SSE data lines asynchronously."""
        base = self._base_url
        url = f"{base}{path}"
        headers = {
            **self._headers,
            "accept": "text/event-stream",
            **(extra_headers or {}),
        }

        async with self._client.stream(
            method,
            url,
            json=json,
            headers=headers,
            timeout=timeout or self._timeout,
        ) as resp:
            if resp.status_code >= 400:
                body_bytes = await resp.aread()
                try:
                    import json as _json

                    body = _json.loads(body_bytes)
                except Exception:
                    body = {"error": {"code": "unknown", "message": body_bytes.decode(errors="replace")}}
                raise _map_error(resp.status_code, body)

            async for line in resp.aiter_lines():
                if line.startswith("data: "):
                    data = line[6:]
                    if data == "[DONE]":
                        return
                    yield data

    async def upload(
        self,
        path: str,
        file_data: bytes,
        filename: str,
        *,
        timeout: float | None = None,
    ) -> dict[str, Any]:
        url = f"{self._platform_base_url}{path}"
        headers = {k: v for k, v in self._headers.items() if k != "content-type"}

        resp = await self._client.post(
            url,
            files={"file": (filename, file_data)},
            headers=headers,
            timeout=timeout or self._timeout,
        )
        if resp.status_code >= 400:
            try:
                body = resp.json()
            except Exception:
                body = {"error": {"code": "unknown", "message": resp.text}}
            raise _map_error(resp.status_code, body)
        return resp.json()

    async def close(self) -> None:
        await self._client.aclose()

File: typescript/src/runtime/transport.ts
/** HTTP transport for AceDataCloud SDK. Uses native fetch (Node 18+). */

import {
  APIError,
  AuthenticationError,
  InsufficientBalanceError,
  ModerationError,
  RateLimitError,
  ResourceDisabledError,
  TimeoutError,
  TokenMismatchError,
  TransportError,
  ValidationError,
} from './errors';
import type {
  PaymentHandler,
  PaymentRequiredBody,
} from './payment';

const ERROR_CODE_MAP: Record<string, typeof APIError> = {
  invalid_token: AuthenticationError,
  token_expired: AuthenticationError,
  no_token: AuthenticationError,
  token_mismatched: TokenMismatchError,
  used_up: InsufficientBalanceError,
  disabled: ResourceDisabledError,
  too_many_requests: RateLimitError,
  bad_request: ValidationError,
};

const RETRY_STATUS_CODES = new Set([408, 409, 429, 500, 502, 503, 504]);

export function mapError(statusCode: number, body: Record<string, unknown>): APIError {
  // The server occasionally returns ``error`` as a bare string (e.g. x402
  // facilitator diagnostics) or omits/nulls it entirely. Normalize to an
  // object so downstream property access is always safe and the message
  // is preserved verbatim.
  let errorData: Record<string, unknown>;
  if (typeof body.error === 'string') {
    errorData = { message: body.error };
  } else if (body.error && typeof body.error === 'object' && !Array.isArray(body.error)) {
    errorData = body.error as Record<string, unknown>;
  } else {
    errorData = {};
  }
  const code = (errorData.code ?? '') as string;
  const message = (errorData.message ?? '') as string;
  const traceId = body.trace_id as string | undefined;

  let ErrorClass = ERROR_CODE_MAP[code];
  if (!ErrorClass) {
    if (statusCode === 403) ErrorClass = ModerationError;
    else if (statusCode === 401) ErrorClass = AuthenticationError;
    else if (statusCode === 429) ErrorClass = RateLimitError;
    else if (statusCode === 400) ErrorClass = ValidationError;
    else ErrorClass = APIError;
  }

  return new ErrorClass({ message, statusCode, code, traceId, body });
}

function backoffDelay(attempt: number): number {
  const base = Math.min(2 ** attempt, 8);
  return base + Math.random() * 0.5;
}

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

function isAbortError(error: unknown): boolean {
  return typeof error === 'object' && error !== null && 'name' in error && error.name === 'AbortError';
}

function timeoutError(error: unknown): TimeoutError {
  return new TimeoutError({
    message: `Request timed out${error instanceof Error && error.message ? `: ${error.message}` : ''}`,
    statusCode: 0,
    code: 'timeout',
  });
}

export interface TransportOptions {
  apiToken?: string;
  baseURL?: string;
  platformBaseURL?: string;
  timeout?: number;
  maxRetries?: number;
  headers?: Record<string, string>;
  /**
   * Optional handler invoked when a request returns `402 Payment Required`.
   * The handler receives the parsed `accepts` list and must return the extra
   * headers (typically `X-Payment`) to attach to the automatic retry.
   *
   * See `@acedatacloud/x402-client` for a drop-in implementation.
   */
  paymentHandler?: PaymentHandler;
}

export class Transport {
  private baseURL: string;
  private platformBaseURL: string;
  private timeout: number;
  private maxRetries: number;
  private headers: Record<string, string>;
  private paymentHandler?: PaymentHandler;

  constructor(opts: TransportOptions = {}) {
    const token = opts.apiToken ?? process.env.ACEDATACLOUD_API_TOKEN ?? '';
    if (!token && !opts.paymentHandler) {
      throw new AuthenticationError({
        message:
          'apiToken is required (or provide a paymentHandler, e.g. from @acedatacloud/x402-client). ' +
          'Pass it to the client or set ACEDATACLOUD_API_TOKEN.',
        statusCode: 0,
        code: 'no_token',
      });
    }
    this.baseURL = (opts.baseURL ?? 'https://x402.acedata.cloud').replace(/\/+$/, '');
    this.platformBaseURL = (opts.platformBaseURL ?? 'https://platform.acedata.cloud').replace(/\/+$/, '');
    this.timeout = opts.timeout ?? 300_000;
    this.maxRetries = opts.maxRetries ?? 2;
    this.paymentHandler = opts.paymentHandler;
    const baseHeaders: Record<string, string> = {
      accept: 'application/json',
      'content-type': 'application/json',
      'user-agent': 'acedatacloud-node/0.1.0',
      ...(opts.headers ?? {}),
    };
    if (token) {
      baseHeaders.authorization = `Bearer ${token}`;
    }
    this.headers = baseHeaders;
  }

  async request(
    method: string,
    path: string,
    opts: {
      json?: Record<string, unknown>;
      params?: Record<string, string>;
      platform?: boolean;
      timeout?: number;
      headers?: Record<string, string>;
    } = {}
  ): Promise<Record<string, unknown>> {
    const base = opts.platform ? this.platformBaseURL : this.baseURL;
    let url = `${base}${path}`;
    if (opts.params) {
      const qs = new URLSearchParams(opts.params).toString();
      url += `?${qs}`;
    }
    const headers = { ...this.headers, ...(opts.headers ?? {}) };
    const timeoutMs = opts.timeout ?? this.timeout;

    let lastError: Error | null = null;
    let paymentAttempted = false;
    let extraHeaders: Record<string, string> = {};
    let attempt = 0;
    while (true) {
      const controller = new AbortController();
      const timer = setTimeout(() => controller.abort(), timeoutMs);
      try {
        const resp = await fetch(url, {
          method,
          headers: { ...headers, ...extraHeaders },
          body: opts.json ? JSON.stringify(opts.json) : undefined,
          signal: controller.signal,
        });
        clearTimeout(timer);

        if (resp.status === 402 && this.paymentHandler && !paymentAttempted) {
          const text = await resp.text();
          let body: unknown;
          try {
            body = JSON.parse(text);
          } catch {
            throw mapError(402, { error: { code: 'invalid_402', message: text } });
          }
          if (
            !body ||
            typeof body !== 'object' ||
            !Array.isArray((body as PaymentRequiredBody).accepts) ||
            !(body as PaymentRequiredBody).accepts.length ||
            !(body as PaymentRequiredBody).accepts.every(
              (requirement) => requirement !== null && typeof requirement === 'object'
            )
          ) {
            throw mapError(402, { error: { code: 'invalid_402', message: 'No payment requirements' } });
          }
          const paymentRequired = body as PaymentRequiredBody;
          const result = await this.paymentHandler({
            url,
            method,
            body: opts.json,
            accepts: paymentRequired.accepts,
          });
          extraHeaders = { ...extraHeaders, ...result.headers };
          paymentAttempted = true;
          continue;
        }

        if (resp.status >= 400) {
          const text = await resp.text();
          let body: Record<string, unknown>;
          try {
            body = JSON.parse(text) as Record<string, unknown>;
          } catch {
            body = { error: { code: 'unknown', message: text } };
          }
          if (!paymentAttempted && RETRY_STATUS_CODES.has(resp.status) && attempt < this.maxRetries) {
            await sleep(backoffDelay(attempt) * 1000);
            attempt += 1;
            continue;
          }
          throw mapError(resp.status, body);
        }

        return (await resp.json()) as Record<string, unknown>;
      } catch (err) {
        clearTimeout(timer);
        if (err instanceof APIError) throw err;
        if (isAbortError(err)) {
          lastError = timeoutError(err);
        } else {
          lastError = err as Error;
        }
        if (!paymentAttempted && attempt < this.maxRetries) {
          await sleep(backoffDelay(attempt) * 1000);
          attempt += 1;
          continue;
        }
        throw lastError;
      }
    }
  }

  async *requestStream(
    method: string,
    path: string,
    opts: { json?: Record<string, unknown>; timeout?: number } = {}
  ): AsyncGenerator<string, void, unknown> {
    const url = `${this.baseURL}${path}`;
    const headers = { ...this.headers, accept: 'text/event-stream' };
    let paymentAttempted = false;
    let extraHeaders: Record<string, string> = {};

    while (true) {
      const controller = new AbortController();
      const timeoutMs = opts.timeout ?? this.timeout;
      const connectionTimer = setTimeout(() => controller.abort(), timeoutMs);
      try {
        let resp: Response;
        try {
          resp = await fetch(url, {
            method,
            headers: { ...headers, ...extraHeaders },
            body: opts.json ? JSON.stringify(opts.json) : undefined,
            signal: controller.signal,
          });
        } catch (error) {
          if (isAbortError(error)) throw timeoutError(error);
          throw error;
        } finally {
          clearTimeout(connectionTimer);
        }

        if (resp.status === 402 && this.paymentHandler && !paymentAttempted) {
          const text = await resp.text();
          let body: unknown;
          try {
            body = JSON.parse(text);
          } catch {
            throw mapError(402, { error: { code: 'invalid_402', message: text } });
          }
          if (
            !body ||
            typeof body !== 'object' ||
            !Array.isArray((body as PaymentRequiredBody).accepts) ||
            !(body as PaymentRequiredBody).accepts.length ||
            !(body as PaymentRequiredBody).accepts.every(
              (requirement) => requirement !== null && typeof requirement === 'object'
            )
          ) {
            throw mapError(402, { error: { code: 'invalid_402', message: 'No payment requirements' } });
          }
          const paymentRequired = body as PaymentRequiredBody;
          const result = await this.paymentHandler({
            url,
            method,
            body: opts.json,
            accepts: paymentRequired.accepts,
          });
          extraHeaders = { ...extraHeaders, ...result.headers };
          paymentAttempted = true;
          continue;
        }

        if (resp.status >= 400) {
          const text = await resp.text();
          let body: Record<string, unknown>;
          try {
            body = JSON.parse(text) as Record<string, unknown>;
          } catch {
            body = { error: { code: 'unknown', message: text } };
          }
          throw mapError(resp.status, body);
        }

        if (!resp.body) throw new TransportError('No response body for stream');

        const reader = resp.body.getReader();
        const decoder = new TextDecoder();
        let buffer = '';

        try {
          while (true) {
            const idleTimer = setTimeout(() => controller.abort(), timeoutMs);
            let result: Awaited<ReturnType<typeof reader.read>>;
            try {
              result = await reader.read();
            } catch (error) {
              if (isAbortError(error)) throw timeoutError(error);
              throw error;
            } finally {
              clearTimeout(idleTimer);
            }
            const { done, value } = result;
            if (done) break;
            buffer += decoder.decode(value, { stream: true });

            const lines = buffer.split('\n');
            buffer = lines.pop() ?? '';

            for (const line of lines) {
              if (line.startsWith('data: ')) {
                const data = line.slice(6);
                if (data === '[DONE]') return;
                yield data;
              }
            }
          }
        } finally {
          try {
            await reader.cancel();
          } catch {
            // Preserve the original stream error, especially SDK TimeoutError.
          }
        }
        return;
      } finally {
        clearTimeout(connectionTimer);
      }
    }
  }

  async upload(
    path: string,
    fileData: Buffer | Uint8Array,
    filename: string,
    opts: { timeout?: number } = {}
  ): Promise<Record<string, unknown>> {
    const url = `${this.platformBaseURL}${path}`;
    const boundary = `----AceDataCloudBoundary${Date.now()}`;
    const headers = {
      ...this.headers,
      'content-type': `multipart/form-data; boundary=${boundary}`,
    };
    delete (headers as Record<string, string>)['content-type'];

    const body = new FormData();
    body.append('file', new Blob([fileData]), filename);

    const authHeaders: Record<string, string> = {
      authorization: this.headers.authorization,
      'user-agent': this.headers['user-agent'],
    };

    const controller = new AbortController();
    const timer = setTimeout(() => controller.abort(), opts.timeout ?? this.timeout);
    try {
      let resp: Response;
      try {
        resp = await fetch(url, {
          method: 'POST',
          headers: authHeaders,
          body,
          signal: controller.signal,
        });
      } catch (error) {
        if (isAbortError(error)) throw timeoutError(error);
        throw error;
      }
      clearTimeout(timer);

      if (resp.status >= 400) {
        const text = await resp.text();
        let respBody: Record<string, unknown>;
        try {
          respBody = JSON.parse(text) as Record<string, unknown>;
        } catch {
          respBody = { error: { code: 'unknown', message: text } };
        }
        throw mapError(resp.status, respBody);
      }
      return (await resp.json()) as Record<string, unknown>;
    } finally {
      clearTimeout(timer);
    }
  }
}
